test/functional/backwards_compatibility.py

```diff
@@ -0,0 +1,212 @@
+        assert_equal(self.nodes[V0_15].getaddressesbyaccount("legacy"), [address1]) 
+        assert_equal(self.nodes[V0_15].getaddressesbyaccount("p2sh"), [address2]) 
+        
+        # bech32 address is corrupted until the next upgrade:
+        assert_equal(self.nodes[V0_15].getaddressesbyaccount("bech32"), ["3QJmnh"])
+        
+        # Scenario 1: create and fund address on node 1 before we override with backup
+        address5 = self.nodes[1].getnewaddress(account="p2sh-post-backup", address_type='p2sh-segwit')
+        address6 = self.nodes[1].getnewaddress(account="bech32-post-backup", address_type='bech32')
+        address7 = self.nodes[1].getnewaddress(account="p2sh-disappear", address_type='p2sh-segwit')
+        self.nodes[0].sendtoaddress(address5, 1)
+        self.nodes[0].sendtoaddress(address6, 1)
+        self.nodes[0].sendtoaddress(address7, 1)
+        self.nodes[0].generate(1)
+        sync_blocks(self.nodes)
+        assert_equal(self.nodes[1].getbalance("p2sh-post-backup"), 1)
+        assert_equal(self.nodes[1].getbalance("bech32-post-backup"), 1)
+        assert_equal(self.nodes[1].getbalance("p2sh-disappear"), 1)
+
+        # Upgrade back to v0.16 by copying to node1:
+        # TODO: use v0.16 node after release
+        self.copy_wallet(V0_15,1)
+
+        assert_equal(self.nodes[1].getbalance("legacy"), 1)
+        assert_equal(self.nodes[1].getbalance("p2sh"), 1) # Scenerio 5
+
+        # Scenario 2
+        assert_equal(self.nodes[1].getaddressesbyaccount("p2sh-nocoin"), [address3])
+        
+        # bech32 account should show funds again (scenerio 2):
+        assert_equal(self.nodes[1].getbalance("bech32"), 1) 
+        
+        # bech32 address format should be restored:
+        assert_equal(self.nodes[1].getaddressesbyaccount("bech32"), [address4])
+        
+        # Check scenario 1, SegWit addresses should come back, even if not explictly requested:
+        self.nodes[1].getnewaddress(account="p2sh-post-backup", address_type='p2sh-segwit')
+        assert_equal(self.nodes[1].getaddressesbyaccount("p2sh-post-backup"), [address5])
+        self.nodes[1].getnewaddress(account="bech32-post-backup", address_type='bech32')
+        assert_equal(self.nodes[1].getaddressesbyaccount("bech32-post-backup"), [address6])
+        
+        self.nodes[1].getnewaddress(account="p2sh-disappear", address_type='bech32') # wrong type!
+        self.nodes[1].getnewaddress(account="p2sh-disappear", address_type='p2sh-segwit') # next key...
+        assert_not_equal(self.nodes[1].getaddressesbyaccount("p2sh-disappear"), [address7])
+        assert_equal(self.nodes[1].getbalance("p2sh-disappear"), 0)
+
+        # Balance won't immedidately update:
+        self.nodes[1].rescanblockchain()
+        assert_equal(self.nodes[1].getbalance("p2sh-post-backup"), 1)
+        assert_equal(self.nodes[1].getbalance("bech32-post-backup"), 1) 
+        assert_equal(self.nodes[1].getbalance("p2sh-disappear"), 0)
```

Comment:
@sipa I'm not sure how to interpret scenario 1. It seems that, for keys generated after a backup was made, funds are irretrievable if the user doesn't create the correct address type for each new key. Is this a bug?